# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1711492787 -3600
# Parent  1807d6b561b470e60f6896e9b6c38f52b82d8735
Feature: limit result to 1000 URLs in ScanTXT(). Was part of ...

  Bug 1812386 - Optimisation as per profiler showing a lot of time spent in FindURLStart/End in nsTSubstring.
  Also limit result to 1000 URLs and match look-ahead to 2000 characters.

diff --git a/netwerk/streamconv/converters/mozTXTToHTMLConv.cpp b/netwerk/streamconv/converters/mozTXTToHTMLConv.cpp
--- a/netwerk/streamconv/converters/mozTXTToHTMLConv.cpp
+++ b/netwerk/streamconv/converters/mozTXTToHTMLConv.cpp
@@ -981,8 +981,9 @@ mozTXTToHTMLConv::ScanTXT(const nsAStrin
     return NS_ERROR_OUT_OF_MEMORY;
   }
 
   bool doURLs = 0 != (whattodo & kURLs);
+  uint32_t countURLs = 0;  // Doing URLs is expensive.
   bool doGlyphSubstitution = 0 != (whattodo & kGlyphSubstitution);
   bool doStructPhrase = 0 != (whattodo & kStructPhrase);
 
   uint32_t structPhrase_strong = 0;  // Number of currently open tags
@@ -1085,8 +1086,10 @@ mozTXTToHTMLConv::ScanTXT(const nsAStrin
                              replaceBefore);
               aOutString += outputHTML;
               endOfLastURLOutput = aOutString.Length();
               i = *ci.Seek(i + replaceAfter);
+              countURLs++;
+              if (countURLs > 1000) doURLs = false;
               continue;
             }
           }
           break;
